### DISCUSSION TITLE
at_before_move on character not working as expected

### DISCUSSION THREAD

**Discussion by pruuk** [evennia/evennia#2508] [Evennia Questions] [answered]
I have the following on my character class:

    # prevent movement into a room if the room is full. This is done using an
    # encumberance trait counter on rooms, just like the one on CharacterCmdSet
    # prevent moves if not standing or if we're out of stamina
    def at_before_move(self, destination):
        """
        Called just before trying to move.
        """
        # check to see if room is full
        if destination.traits.enc.current + self.traits.mass.actual > destination.traits.enc.max:
            # room is full, cancel move
            self.msg("That room has too many things and/or people in it. You can't enter right now.")
            return False
        else:
            return True
        # check to see if we're standing before we try to move
        if self.db.info['Position'] != 'standing':
            self.msg("You need to stand to move around.")
            return False
        else:
            return True
        # check to see if we're out of stamina. If we are, prevent move
        if self.traits.sp.current < 10:
            self.msg("You don't have the stamina to travel right now. Rest up and try again later.")
            self.exhaustion_check()
            return False
        else:
            return True

I've examined my test player:

Persistent attribute(s):
 info = {'Target': Steve, 'Mercy': True, 'Default Attack': 'melee_weapon_strike', 'In Combat': False, 'Position': 'resting', 'Sneaking': True, 'Wimpy': 100, 'Yield': 200, 'Title': None}

Even though my position is 'resting' instead of standing, this test character is still able to move from room to room. What am I missing?

**Comment by kovitikus** [comment, marked as answer]
You never check if you are standing or not. You `return False` if the comparison line is `True`, and you `return True` no matter what.

`return` means that it breaks away from the method and goes back to what it was doing before this method was called, therefore nothing past the `else: return True` ever gets read by Python.

What you can do is set a boolean variable that you set in these conditions.

```py
# prevent movement into a room if the room is full. This is done using an
# encumberance trait counter on rooms, just like the one on CharacterCmdSet
# prevent moves if not standing or if we're out of stamina
def at_before_move(self, destination):
    """
    Called just before trying to move.
    """
    can_move = False

    # check to see if room is full
    if destination.traits.enc.current + self.traits.mass.actual > destination.traits.enc.max:
        # room is full, cancel move
        self.msg("That room has too many things and/or people in it. You can't enter right now.")
        return can_move
    else:
        can_move = True

    # check to see if we're standing before we try to move
    if self.db.info['Position'] != 'standing':
        self.msg("You need to stand to move around.")
        return can_move
    else:
        can_move = True

    # check to see if we're out of stamina. If we are, prevent move
    if self.traits.sp.current < 10:
        self.msg("You don't have the stamina to travel right now. Rest up and try again later.")
        self.exhaustion_check()
        return can_move
    else:
        can_move = True

    return can_move
```

Yet another way to do this, with even less lines, is to only return or set anything if it's false, until the very end.

```py
# prevent movement into a room if the room is full. This is done using an
# encumberance trait counter on rooms, just like the one on CharacterCmdSet
# prevent moves if not standing or if we're out of stamina
def at_before_move(self, destination):
    """
    Called just before trying to move.
    """
    # check to see if room is full
    if destination.traits.enc.current + self.traits.mass.actual > destination.traits.enc.max:
        # room is full, cancel move
        self.msg("That room has too many things and/or people in it. You can't enter right now.")
        return False

    # check to see if we're standing before we try to move
    if self.db.info['Position'] != 'standing':
        self.msg("You need to stand to move around.")
        return False

    # check to see if we're out of stamina. If we are, prevent move
    if self.traits.sp.current < 10:
        self.msg("You don't have the stamina to travel right now. Rest up and try again later.")
        self.exhaustion_check()
        return False

    return True
```

The method will only reach the bottom and `return True` if none of the other conditions were `True`. But, the second any of them are `True`, it breaks away and sends back `False`.

I highly recommend using the second version I posted, as it is not only shorter, but easier to read the logic and you only need one line to `return True` back to the code that will be checking this boolean.

**Reply by pruuk** [reply]
Thanks, I knew I was missing something incredibly obvious. :)

**Reply by kovitikus** [reply]
No problem. 😎
